### PR TITLE
adding otel capabilities

### PR DESCRIPTION
**Pull request by asteurer**
With this PR, a user will be able to define certain OTel parameters (which we pre-approve) that will allow them to view trace information via opentelemetry-collector and jaeger.

### CODE REVIEW

**Comment by asteurer** [comment]
Is there a place where I can leave instructions for how to deploy jaeger and opentelemetry-collector properly for SpinKube?

**Comment by kate-goldenring** [comment]
> Is there a place where I can leave instructions for how to deploy jaeger and opentelemetry-collector properly for SpinKube?

I'd add it here https://github.com/spinkube/documentation/tree/main/content/en/docs/topics

File: config/samples/hpa.yaml [hunk 1]
```diff
@@ -32,3 +32,4 @@ spec:
       target:
         type: Utilization
         averageUtilization: 50
+
```
Comment (calebschoepp) [inline config/samples/hpa.yaml:35]: Nit
```suggestion

```

Comment (calebschoepp) [reply config/samples/hpa.yaml:35]: This probably shouldn't be modified.

File: internal/controller/deployment.go [hunk 2]
```diff
@@ -121,6 +118,15 @@ func ConstructEnvForApp(ctx context.Context, app *spinv1alpha1.SpinApp) []corev1
 		envs[idx] = env
 	}
 
+	// Adding the OpenTelemetry params
+	for _, entry := range app.Spec.Otel {
+		envs = append(envs, corev1.EnvVar{Name: entry.Name, Value: entry.Value, ValueFrom: entry.ValueFrom})
+	}
+
+	if len(envs) == 0 {
```
Comment (calebschoepp) [inline internal/controller/deployment.go:126]: Why is this here?

File: api/v1alpha1/spinapp_types.go [hunk 3]
```diff
@@ -80,6 +80,9 @@ type SpinAppSpec struct {
 
 	// Resources defines the resource requirements for this app.
 	Resources Resources `json:"resources,omitempty"`
+
+	// Otel provides Kubernetes Bindings to Otel Variables.
+	Otel []OtelVar `json:"otel,omitempty"`
```
Comment (endocrimes) [inline api/v1alpha1/spinapp_types.go:85]: Rather than being a bunch of loose kvs in a list I'd much rather have this be less flexible and more structured

Like:

type OtelConfig struct {
 Exporter string
}

**Review by endocrimes — CHANGES_REQUESTED** [review]

File: internal/controller/deployment.go (hunk 2 above)
Comment (asteurer) [reply internal/controller/deployment.go:126]: The original buildEnvs had a conditional at the beginning which returned `nil` if there were no spin variables defined. I assumed that returning an empty `[]corev1.EnvVar` would be a problem, so I returned `nil` if neither spin nor otel variables were defined. Sounds like it might not be necessary?

File: api/v1alpha1/spinapp_types.go [hunk 4]
```diff
@@ -80,6 +80,9 @@ type SpinAppSpec struct {
 
 	// Resources defines the resource requirements for this app.
 	Resources Resources `json:"resources,omitempty"`
+
```
Comment (endocrimes) [inline api/v1alpha1/spinapp_types.go:83]: @calebschoepp i thought we were going to put these on the Executor because they're mostly common to a set of apps, not an individual app?

File: api/v1alpha1/spinapp_types.go (hunk 3 above)
Comment (asteurer) [reply api/v1alpha1/spinapp_types.go:85]: So like this: 
```
  otel:
    - OTEL_EXPORTER_OTLP_ENDPOINT: http://otel-collector-opentelemetry-collector.default.svc.cluster.local:4318
```

rather than this: 
```
  otel:
    - name: OTEL_EXPORTER_OTLP_ENDPOINT
      value: http://otel-collector-opentelemetry-collector.default.svc.cluster.local:4318
```

Comment (endocrimes) [reply api/v1alpha1/spinapp_types.go:85]: no like this:

```yaml
otel:
  endpoint: http://otel-collector-opentelemetry-collector.default.svc.cluster.local:4318
```

that we convert into envs/however the shim needs config in the operator

Comment (calebschoepp) [reply api/v1alpha1/spinapp_types.go:85]: More specifically I prefer the bag of kv's approach b/c than we don't have to churn on the CRD, we can churn on the validation in the operator. OTEL has a __lot__ of different configuration options and I really don't want to bake them all into our CRD.

File: api/v1alpha1/spinapp_types.go (hunk 4 above)
Comment (calebschoepp) [reply api/v1alpha1/spinapp_types.go:83]: Yep we were, this just got mixed up in the communcation between Andrew and I.

@asteurer let's move this over to `spinapp_executor_types.go`. This will change how you'll have to pull the value in the operator.

File: internal/controller/deployment.go (hunk 2 above)
Comment (calebschoepp) [reply internal/controller/deployment.go:126]: Not sure if it is necessary, but your logic is reasonable to me. Let's leave it like this.

File: api/v1alpha1/spinapp_types.go (hunk 3 above)
Comment (endocrimes) [reply api/v1alpha1/spinapp_types.go:85]: Building a stable interface is much cheaper (long term) than adding loosely defined values specific to a single implementation, and our entire design thus far has been to avoid the latter.

Comment (calebschoepp) [reply api/v1alpha1/spinapp_types.go:85]: I want to push back on two things:

1. I expect that there will be very little difference in the ENV keys between executors. The OTEL env vars are standardized for a reason. In the case where an executor does need some very specific env vars it would probably be in its own field outside of the otel field.
2. The loose KV approach is still defining a stable interface. The stable interface just lives at the validation layer in the operator where we validate which keys are allowed.

Do they still have to uninstall/reinstall CRDs when fields are added?

Comment (endocrimes) [reply api/v1alpha1/spinapp_types.go]: > Do they still have to uninstall/reinstall CRDs when fields are added?
Nope

2) The downside there is that it's not a well known, published interface - so offline validation tools and editor autocomplete can never work properly - with a well defined API that comes with an OpenAPI spec, it can.

Comment (calebschoepp) [reply api/v1alpha1/spinapp_types.go]: > The downside there is that it's not a well known, published interface - so offline validation tools and editor autocomplete can never work properly - with a well defined API that comes with an OpenAPI spec, it can.

That is a downside.

> Do they still have to uninstall/reinstall CRDs when fields are added?
> Nope

That changes my mind. How does it work then, do they just apply the new CRD and it patches?

Comment (endocrimes) [reply api/v1alpha1/spinapp_types.go]: Yep!

File: internal/controller/deployment.go [hunk 5]
```diff
@@ -121,6 +118,10 @@ func ConstructEnvForApp(ctx context.Context, app *spinv1alpha1.SpinApp) []corev1
 		envs[idx] = env
 	}
 
+	// Adding the OpenTelemetry params
+	otelVars := []corev1.EnvVar{{Name: "OTEL_EXPORTER_OTLP_ENDPOINT", Value: otel.Endpoint}}
+	envs = append(envs, otelVars...)
```
Comment (endocrimes) [inline internal/controller/deployment.go:123]:
```suggestion
	if otel.Endpoint != "" {
	  envs = append(envs, corev1.EnvVar{Name: "OTEL_EXPORTER_OTLP_ENDPOINT", Value: otel.Endpoint})
	}
```

no need to construct a second list - can just directly add things to envs.

File: internal/webhook/spinappexecutor_validating.go [hunk 6]
```diff
@@ -50,5 +53,21 @@ func (v *SpinAppExecutorValidator) ValidateDelete(ctx context.Context, obj runti
...
+	}
+	if len(allErrs) == 0 {
+		return nil
+	}
+
+	return apierrors.NewInvalid(schema.GroupKind{Group: "core.spinoperator.dev", Kind: "SpinAppExecutor"}, executor.Name, allErrs)
+}
+
+func validateOpenTelemetry(deployment spinv1alpha1.ExecutorDeploymentConfig) *field.Error {
+	if deployment.Otel.Endpoint == "" {
```
Comment (endocrimes) [inline internal/webhook/spinappexecutor_validating.go:68]: Do we actually want to make otel a requirement? - probably should be optional

**Review by endocrimes — CHANGES_REQUESTED** [review]
Mostly lgtm, blocking on the discussion around optional/required, not the env cleanup

File: internal/webhook/spinappexecutor_validating.go (hunk 6 above)
Comment (calebschoepp) [reply internal/webhook/spinappexecutor_validating.go:68]: +1 to it being optional

File: config/samples/hpa.yaml (hunk 1 above)
Comment (calebschoepp) [reply config/samples/hpa.yaml:35]: Looks like the file is still modified for some reason.

File: config/samples/otel.yaml [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+  name: containerd-shim-spin
+spec:
+  createDeployment: true
+  deploymentConfig:
+    runtimeClassName: wasmtime-spin-v2
+    installDefaultCACerts: true
+    otel:
+      endpoint: http://otel-collector-opentelemetry-collector.default.svc.cluster.local:4318
+
+
```
Comment (calebschoepp) [inline config/samples/otel.yaml:31]: Nit: I like to make the samples as simple and as relevant to what you're demoing as possible.

```suggestion
apiVersion: core.spinoperator.dev/v1alpha1
kind: SpinApp
metadata:
  name: otel-spinapp
spec:
  image: ghcr.io/spinkube/spin-operator/cpu-load-gen:20240311-163328-g1121986
  executor: containerd-shim-spin
  replicas: 1
---
apiVersion: core.spinoperator.dev/v1alpha1
kind: SpinAppExecutor
metadata:
  name: containerd-shim-spin
spec:
  createDeployment: true
  deploymentConfig:
    runtimeClassName: wasmtime-spin-v2
    installDefaultCACerts: true
    otel:
      endpoint: http://otel-collector-opentelemetry-collector.default.svc.cluster.local:4318
```

File: internal/controller/deployment_test.go [hunk 8]
```diff
@@ -142,7 +142,7 @@ func TestConstructEnvForApp(t *testing.T) {
 				},
 			}
 
-			envs := ConstructEnvForApp(context.Background(), app)
+			envs := ConstructEnvForApp(context.Background(), app, spinv1alpha1.OtelConfig{})
```
Comment (calebschoepp) [inline internal/controller/deployment_test.go:145]: Maybe set something in the `OtelConfig` to actually test it is being set properly?

**Comment by asteurer** [comment]
Following up on this

File: config/manager/kustomization.yml [hunk 9]
```diff
@@ -1,2 +1,8 @@
 resources:
-- manager.yaml
\ No newline at end of file
+- manager.yaml
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+images:
+- name: ghcr.io/spinkube/spin-operator
+  newName: ttl.sh/new-spin-operator
+  newTag: 1h
```
[the remark this answers is not in the document]
Comment (calebschoepp) [reply config/manager/kustomization.yml:8]: This snuck back in somehow.

File: config/samples/hpa.yaml [hunk 10]
```diff
@@ -31,4 +31,4 @@ spec:
       name: cpu
       target:
         type: Utilization
-        averageUtilization: 50
+        averageUtilization: 50
```
Comment (calebschoepp) [inline config/samples/hpa.yaml:34]: This change isn't related to this PR and should not be in this diff.

File: config/samples/otel.yaml [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
```
Comment (calebschoepp) [inline config/samples/otel.yaml:22]: Nit: Some extra newline stuff going on here. Not super important (hence the nit), but I'm overly sensitive to this kind of stuff 🤣. Ultimately whitespace doesn't matter, but I find clean whitespace makes code easier to read and I also find that clean whitespace shows craftsmanship and care for your code. 🤷 just my 2¢.

File: config/samples/otel.yaml [hunk 12]
```diff
@@ -0,0 +1,22 @@
+apiVersion: core.spinoperator.dev/v1alpha1
+kind: SpinApp
+metadata:
+  name: otel-spinapp
+spec:
+  image: ghcr.io/spinkube/spin-operator/cpu-load-gen:20240311-163328-g1121986
+  executor: containerd-shim-spin
+  replicas: 1
+---
+apiVersion: core.spinoperator.dev/v1alpha1
```
Comment (calebschoepp) [inline config/samples/otel.yaml:10]: I feel like in the documentation we typically have them already install this `containerd-shim-spin` executor. I worry that if they just `k apply -f config/samples/otel.yaml` that they will get an error because it already exists. What if you made this a different executor like `my-custom-otel-shim-executor` (obvs choose a better name).

File: config/samples/otel.yaml [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+kind: SpinAppExecutor
+metadata:
+  name: containerd-shim-spin
+spec:
+  createDeployment: true
+  deploymentConfig:
+    runtimeClassName: wasmtime-spin-v2
+    installDefaultCACerts: true
+    otel:
+      endpoint: http://otel-collector-opentelemetry-collector.default.svc.cluster.local:4318
```
Comment (calebschoepp) [inline config/samples/otel.yaml:20]: How would the user have this running already? Can we document that somewhere? Or alternatively can we have this sample yaml also deploy the collector? That might be hard b/c I think you need to use helm.

File: internal/controller/deployment_test.go [hunk 14]
```diff
@@ -94,6 +94,9 @@ func TestConstructEnvForApp(t *testing.T) {
 
 		value     string
 		valueFrom *corev1.EnvVarSource
+
+		expectedOtelEndpointVarName string
+		otelEndpointValue           string
```
Comment (calebschoepp) [inline internal/controller/deployment_test.go:99]: I wonder if there is a more generic way you can test this? This isn't going to scale very well the second we start adding more than a few otel vars (of which there will be many).

Comment (calebschoepp) [reply internal/controller/deployment_test.go:99]: Perhaps just pass an otel values struct and then a map of the expected var name to value mapping?

File: internal/webhook/spinappexecutor_validating.go [hunk 15]
```diff
@@ -50,5 +53,24 @@ func (v *SpinAppExecutorValidator) ValidateDelete(ctx context.Context, obj runti
...
+	}
+
+	return apierrors.NewInvalid(schema.GroupKind{Group: "core.spinoperator.dev", Kind: "SpinAppExecutor"}, executor.Name, allErrs)
+}
+
+func validateOpenTelemetry(deployment spinv1alpha1.ExecutorDeploymentConfig) *field.Error {
+	// Only performing validations if the OTel field is initialized
+	if deployment.Otel != nil {
+		if deployment.Otel.Endpoint == "" {
+			return field.NotFound(field.NewPath("otel"), deployment.Otel.Endpoint)
```
Comment (calebschoepp) [inline internal/webhook/spinappexecutor_validating.go:71]: Is this just validating it isn't empty? Now that we're baking the otel vars into the schema I wonder if it makes sense to just not have this validation logic.

File: api/v1alpha1/spinappexecutor_types.go [hunk 16]
```diff
@@ -56,6 +59,11 @@ type SpinAppExecutorStatus struct {
 	// Important: Run "make" to regenerate code after modifying this file
 }
 
+// OtelConfig is the supported environment variables for OpenTelemetry
+type OtelConfig struct {
+	Endpoint string `json:"endpoint"`
+}
+
```
Comment (calebschoepp) [inline api/v1alpha1/spinappexecutor_types.go:66]:
```suggestion
type OtelConfig struct {
	// Sets OTEL_EXPORTER_OTLP_ENDPOINT
	OtelExporterOtlpEndpoint string `json:"otel_exporter_otlp_endpoint"`
}

```

I'd personally like to see it look like this. The stutter of the otel is annoying, but I prefer the clarity of knowing exactly which variable you're configuring

**Review by calebschoepp — CHANGES_REQUESTED** [review]

File: config/samples/hpa.yaml (hunk 10 above)
Comment (asteurer) [reply config/samples/hpa.yaml:34]: I'm a little lost...It looks like nothing was changed? How do I remove a diff that effectively hasn't changed anything?

File: config/samples/otel.yaml (hunk 12 above)
Comment (asteurer) [reply config/samples/otel.yaml:10]: How about `otel-shim-executor`?

File: config/samples/hpa.yaml (hunk 10 above)
Comment (calebschoepp) [reply config/samples/hpa.yaml:34]: If you look at the diff you can see a red circle with a line in it. That means there is no trailing newline at the end of the file. The change is that you've removed this trailing newline.

File: config/samples/otel.yaml (hunk 13 above)
Comment (asteurer) [reply config/samples/otel.yaml:20]: Yeah, I think we'd need a helm install. Can we document this using comments inside of the sample app, or is there a place we should document this?

File: internal/webhook/spinappexecutor_validating.go (hunk 15 above)
Comment (asteurer) [reply internal/webhook/spinappexecutor_validating.go:71]: I wanted to have more helpful error messages regarding required fields. Is there a better way to do this?

File: config/samples/otel.yaml (hunk 13 above)
Comment (calebschoepp) [reply config/samples/otel.yaml:20]: Comments inside the yaml seems fine to me. A good followup would be adding more explicit documentation to <https://spinkube.dev/docs>.

File: internal/webhook/spinappexecutor_validating.go (hunk 15 above)
Comment (calebschoepp) [reply internal/webhook/spinappexecutor_validating.go:71]: You could directly [mark things](https://book.kubebuilder.io/reference/markers) as optional or required in the CRD definition. But, I'm not really convinced we want to do that. I think it is hard for us to know in advance how they want to configure it e.g. what if they only want to set `OTEL_EXPORTER_OTLP_TRACES_ENDPOINT` and not `OTEL_EXPORTER_OTLP_ENDPOINT`?

Comment (endocrimes) [reply internal/webhook/spinappexecutor_validating.go:71]: I'd be pro dropping manual validation for now and ratcheting down later in this case 👍

File: api/v1alpha1/spinappexecutor_types.go (hunk 16 above)
Comment (endocrimes) [reply api/v1alpha1/spinappexecutor_types.go:66]: I don't think that _feels_ particularly idiomatic (and is a lot of bonus typing):

```yaml
otel:
  otel_exporter_otlp_endpoint: "mysvc.com"
```

File: api/v1alpha1/spinappexecutor_types.go [hunk 17]
```diff
@@ -61,7 +61,20 @@ type SpinAppExecutorStatus struct {
 
 // OtelConfig is the supported environment variables for OpenTelemetry
 type OtelConfig struct {
-	Endpoint string `json:"endpoint"`
+	// Sets OTEL_EXPORTER_OTLP_ENDPOINT environment variable
```
Comment (endocrimes) [inline api/v1alpha1/spinappexecutor_types.go:64]: I still feel like:

```yaml
otel:
  otel_exporter_*
```

feels really uncomfortable to actually write, and i'm not sure what benefit there is to it - @calebschoepp can you elaborate on why you'd prefer this?

File: api/v1alpha1/spinappexecutor_types.go [hunk 18]
```diff
@@ -61,7 +61,20 @@ type SpinAppExecutorStatus struct {
...
+	OtelExporterOtlpTracesEndpoint string `json:"otel_exporter_otlp_traces_endpoint"`
+
+	// Sets OTEL_EXPORTER_OTLP_METRICS_ENDPOINT environment variable
+	OtelExporterOtlpMetricsEndpoint string `json:"otel_exporter_otlp_metrics_endpoint"`
+
+	// Sets OTEL_EXPORTER_OTLP_LOGS_ENDPOINT environment variable
+	OtelExporterOtlpLogsEndpoint string `json:"otel_exporter_otlp_logs_endpoint"`
+
+	// Sets OTEL_SERVICE_NAME environment variable
+	OtelServiceName string `json:"otel_service_name"`
```
Comment (endocrimes) [inline api/v1alpha1/spinappexecutor_types.go:77]: we probably shouldn't attach a ServiceName to the executor - I assumed we were gonna populate that with app names

Comment (asteurer) [reply api/v1alpha1/spinappexecutor_types.go:77]: @calebschoepp had me add this. Maybe he can elaborate?

Comment (calebschoepp) [reply api/v1alpha1/spinappexecutor_types.go:77]: We don't yet have the functionality to populate service name based off app names. I also don't think that feature is trivial b/c most open telemetry libraries are very good about allowing you to dynamically set the resource. This env var is a workaround for that reality.

@endocrimes if you still feel strongly that we don't have this I'm fine to just remove it. I want to get this PR in ASAP.

File: api/v1alpha1/spinappexecutor_types.go (hunk 17 above)
Comment (calebschoepp) [reply api/v1alpha1/spinappexecutor_types.go:64]: I like the clarity of spelling out the full env var your setting.

I'd be okay to remove the `otel_` prefix b/c it seems like all OTel env vars have that, but I wouldn't want to remove anything else for the sake of clarity.

File: api/v1alpha1/spinappexecutor_types.go (hunk 18 above)
Comment (endocrimes) [reply api/v1alpha1/spinappexecutor_types.go:77]: I think we should think this through - so drop from this PR

File: api/v1alpha1/spinappexecutor_types.go (hunk 17 above)
Comment (endocrimes) [reply api/v1alpha1/spinappexecutor_types.go:64]: dropping the `otel_` prefix at least deals with the stutter so that works

File: api/v1alpha1/spinappexecutor_types.go [hunk 19]
```diff
@@ -56,6 +59,21 @@ type SpinAppExecutorStatus struct {
...
+	OtelExporterOtlpEndpoint string `json:"exporter_otlp_endpoint"`
+
+	// Sets OTEL_EXPORTER_OTLP_TRACES_ENDPOINT environment variable
+	OtelExporterOtlpTracesEndpoint string `json:"exporter_otlp_traces_endpoint"`
+
+	// Sets OTEL_EXPORTER_OTLP_METRICS_ENDPOINT environment variable
+	OtelExporterOtlpMetricsEndpoint string `json:"exporter_otlp_metrics_endpoint"`
+
+	// Sets OTEL_EXPORTER_OTLP_LOGS_ENDPOINT environment variable
+	OtelExporterOtlpLogsEndpoint string `json:"exporter_otlp_logs_endpoint"`
```
Comment (endocrimes) [inline api/v1alpha1/spinappexecutor_types.go:74]:
```suggestion
	// ExporterOtlpEndpoint configures the default combined otlp endpoint for sending telemetry
	ExporterOtlpEndpoint string `json:"exporter_otlp_endpoint,omitempty"`

	// ExporterOtlpTracesEndpoint configures the trace-specific otlp endpoint
	ExporterOtlpTracesEndpoint string `json:"exporter_otlp_traces_endpoint,omitempty"`

	// ExporterOtlpMetricsEndpoint configures the metrics-specific otlp endpoint
	ExporterOtlpMetricsEndpoint string `json:"exporter_otlp_metrics_endpoint,omitempty"`

	// ExporterOtlpLogsEndpoint configures the logs-specific otlp endpoint
	ExporterOtlpLogsEndpoint string `json:"exporter_otlp_logs_endpoint,omitempty"`
```

this PR lgtm with a smol change to naming here - and updating comments to match go standards (starting the comment with the thing it's attached to's name).

Also needs a new `make manifests` committing because the current crd seems out-of-date.

Thanks for working on this!

**Review by endocrimes — APPROVED** [review]
Needs a rebase but LGTM

**Review by calebschoepp — APPROVED** [review]
